### PR TITLE
Mount block PVC rw in clone source pods

### PR DESCRIPTION
**Pull request by akalenyu**
**What this PR does / why we need it**:
Some CSI drivers [0] don't allow mounting block ro, so they simply cannot perform host assisted clones.
Host assisted clones are important as they are a last resort in some cases like switching over to a new storage provisioner.

This may or may not have some performance implications (parallel cloning from single source)
but we are choosing compatibility over performance in this case.

[0] https://github.com/dell/csi-unity/blob/76cb78decd0c7e0ae3bded9668e7fbe9b39dd2ff/service/node.go#L420-L424

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://issues.redhat.com/browse/CNV-33859

**Special notes for your reviewer**:

**Release note**:

```release-note
BugFix: Some provisioners don't allow mounting block PVC ro
```

### CODE REVIEW

File: pkg/controller/common/util.go [hunk 1]
```diff
@@ -603,6 +603,12 @@ func GetPodsUsingPVCs(ctx context.Context, c client.Client, namespace string, na
 									onlyReadOnly = false
 								}
 							}
+							for _, vm := range c.VolumeDevices {
+								if vm.Name == volume.Name {
+									// Node level rw mount and container can't mount block device ro
+									onlyReadOnly = false
+								}
+							}
```
Comment (akalenyu) [inline pkg/controller/common/util.go:611]: @mhenriks @awels I think this check was always broken for pods with volumeDevices

**Review by awels — APPROVED** [review]

File: pkg/controller/common/util.go [hunk 2]
```diff
@@ -612,6 +618,12 @@ func GetPodsUsingPVCs(ctx context.Context, c client.Client, namespace string, na
 						// all mounts must be ro
 						addPod = false
 					}
+					if strings.HasSuffix(pod.Name, common.ClonerSourcePodNameSuffix) && pod.Labels != nil &&
+						pod.Labels[common.CDIComponentLabel] == common.ClonerSourcePodName {
+						// Host assisted clone source pod only reads from source
+						// But some drivers disallow mounting a block PVC ReadOnly
+						addPod = false
```
Comment (alromeros) [inline pkg/controller/common/util.go:625]: Several consecutive conditions lead to the same result (false addPod), would it be possible to clean-up this a little so we have it more condenssed? Maybe something like a `shouldAddPod` function? Not a must but I think this code is getting hard to follow.

File: pkg/controller/clone-controller_test.go [hunk 3]
```diff
@@ -178,6 +178,41 @@ var _ = Describe("Clone controller reconcile loop", func() {
 		}),
 	)
 
+	It("Should create new source pod if source PVC is in use by other host assisted source pod", func() {
```
Comment (alromeros) [inline pkg/controller/clone-controller_test.go:181]: Nice test

File: pkg/controller/common/util.go (hunk 2 above)
Comment (akalenyu) [reply pkg/controller/common/util.go:625]: Yeah the whole function can be refactored but I prefer to stay on the safe side since this is going to backport all the way to 0.55

**Comment by mhenriks** [comment]
/hold
let me look at this

File: pkg/controller/common/util.go (hunk 1 above)
Comment (mhenriks) [reply pkg/controller/common/util.go:611]: Kind of.  This will only cover cases where block PVCs are in the pod spec read/write but not referred to by a container.

File: pkg/controller/common/util.go [hunk 4]
```diff
@@ -612,6 +618,12 @@ func GetPodsUsingPVCs(ctx context.Context, c client.Client, namespace string, na
 						// all mounts must be ro
 						addPod = false
 					}
+					if strings.HasSuffix(pod.Name, common.ClonerSourcePodNameSuffix) && pod.Labels != nil &&
+						pod.Labels[common.CDIComponentLabel] == common.ClonerSourcePodName {
```
Comment (mhenriks) [inline pkg/controller/common/util.go:622]: `pod.Labels != nil` is not required.   `pod.Labels[common.CDIComponentLabel] == common.ClonerSourcePodName` can still be executed when pod.Labels is nil
